Build working tar from dockerfile

hjaensch7 opened the pull request.

hjaensch7 committed:
Clone repo and build functioning docker tar

hjaensch7 committed:
Split build from tar function and changed image path

Others commented on main.go:
```diff
@@ -7,11 +7,13 @@ import (
-	port = ":3069"
+	path     = "/payload"
+	port     = ":3069"
+	tar_name = "test"
```
Generate this from somewhere instead... Also it's clumsy to have function parameters overload global variable names

hjaensch7 replied:
Are you suggesting a config file or something like that?

Others replied:
No. I think you should generate a somewhat random name. Getting a bunch of tars with the same image name may mess stuff up

hjaensch7 replied:
That makes sense. I'll look into using a combination of the username and reponame.

Others commented on main.go:
```diff
@@ -38,7 +40,7 @@ func setLogStreams(
 
 func main() {
-	//Init Log streams
+	//Setup log streams
```
Space at start of comments?

hjaensch7 replied:
Is that something you want me to change?

Others replied:
What's the style in Go?

hjaensch7 replied:
The style is to have a space, I'll add it

hjaensch7 committed:
Updated tar naming to use github fullname

hjaensch7 committed:
Updated tar naming to use random uuid

hjaensch7 committed:
Build tar from dockerfile activate worker locally

This builds the docker image then converts it to a tar. Then the
local worker is activated and pointed to the tar

hjaensch7 committed:
Build tar zip tar send to worker and activate

This is a large commit that includes all of the scp code

Commit:
AWS Config in progress

hjaensch7 committed:
Updated activate to use github repo name and user

- Moved logging to it's own file
- updated paths to reflect AWS instance execution

hjaensch7 committed:
Clean up tar and temp_dir

Others commented on activate.go:
```diff
@@ -0,0 +1,56 @@
+)
+
+const (
+	worker_url = "http://ec2-54-211-200-158.compute-1.amazonaws.com/meta/activate"
```
Load worker urls from environment variables

```
V9_WORKER_URLS=$WORKER_ONE_URL;$WORKER_TWO_URL;...
```

Others commented on activate.go:
```diff
@@ -0,0 +1,56 @@
+	worker_url = "http://ec2-54-211-200-158.compute-1.amazonaws.com/meta/activate"
+)
+
+type dev_id struct {
```
Types should have consistent style

hjaensch7 replied:
What do you mean by this? These variables are a way for go to build the body of the post request easily, they are in the same structure as the `activate` body.

Others commented on activate.go:
```diff
@@ -0,0 +1,56 @@
+}
+
+// Build activate post body
+func createActivateBody(dev dev_id, tarPath string, execution_method string) ([]byte, error) {
```
Inconsistent syntax

hjaensch7 replied:
I don't understand what is inconsistent about this. The parameters? the return values?

Others commented on activate.go:
```diff
@@ -0,0 +1,56 @@
+		return err
+	}
+
+	defer resp.Body.Close()
```
Spacing

Others commented on log.go:
```diff
@@ -0,0 +1,28 @@
+	"log"
+)
+
+var (
+	Info    *log.Logger
+	Warning *log.Logger
+	Error   *log.Logger
+)
```
Is there a way to do this without globals?

Others commented on log.go:
```diff
@@ -0,0 +1,28 @@
+		log.Ldate|log.Ltime|log.Lshortfile)
+	Error = log.New(infoHandle,
+		"ERROR: ",
+		log.Ldate|log.Ltime|log.Lshortfile)
```
`gofmt` give you this?

hjaensch7 replied:
this is just wrapping around the standard `log` package to provide more info and clearly distinguish error vs warning vs info.

Others commented on main.go:
```diff
@@ -1,44 +1,21 @@
 const (
 	path = "/payload"
-	port = ":3069"
+	port = "0.0.0.0:81"
```
"port" is now a misleading name

Others commented on main.go:
```diff
@@ -69,14 +104,38 @@ func main() {
 	}
```
Doing way way too much in main. Break this up

hjaensch7 replied:
I agree, I began to break up some of it with activate and scp but I'll continue to pull more out

Others commented on main.go:
```diff
@@ -56,10 +33,68 @@ func main() {
-		err = downloadRepo(downloadURL, "./test_repo")
+
+		// Get Repo Contents
+		err = downloadRepo(downloadURL, "./temp_repo")
```
defer deleting the repo here, instead of doing it at the end

Others commented on main.go:
```diff
@@ -56,10 +33,68 @@ func main() {
 		}
+
+		// Get random tar name
+		tar_name := guuid.New()
```
Call `.String` here

Others commented on main.go:
```diff
@@ -56,10 +33,68 @@ func main() {
+		err = buildTarFromImage(tar_name.String())
+		if err != nil {
+			Error.Println("Error building tar from image", err)
+		}
```
No return?

Others commented on main.go:
```diff
@@ -56,10 +33,68 @@ func main() {
+		err = buildImageFromDockerfile(tar_name.String())
+		if err != nil {
+			Error.Println("Error building image from Dockerfile", err)
+		}
```
No return?

Others commented on main.go:
```diff
@@ -56,10 +33,68 @@ func main() {
+		tarNameExt = tarNameExt + ".gz"
+
+		// Send .tar to worker
+		source := "/home/ubuntu/go/src/v9_deployment_manager/" + tarNameExt
```
Do _NOT_ hardcode paths. You must get this programmatically (perhaps just do everything in the working directory with relative paths)

Others commented on main.go:
```diff
@@ -56,10 +33,68 @@ func main() {
+		user := push.Repository.Owner.Login
+		repo := push.Repository.Name
+		dev := dev_id{user, repo, "test_hash"}
+		err = activateWorker(dev, "test", destination, tarNameExt)
```
Do not hardcode the user/repo!

hjaensch7 replied:
This is not hard coded, this is reading the user and repo from the github push event

Others commented on main.go:
```diff
@@ -56,10 +33,68 @@ func main() {
+		err = os.RemoveAll("./temp_repo")
+		if err != nil {
+			Error.Println("Failed to remove temp_repo")
+		}
```
Should be defered above, also missing a return?

Others commented on main.go:
```diff
@@ -56,10 +33,68 @@ func main() {
+		err = os.Remove("./" + tarNameExt)
+		if err != nil {
+			Error.Println("Failed to remove tar")
+		}
```
See above

Others commented on scp.go:
```diff
@@ -0,0 +1,49 @@
+func scpToWorker(source string, dest string, tarName string) error {
+	// Use SSH key authentication from the auth package
+	// we ignore the host key in this example, please change this if you use this library
+	clientConfig, err := auth.PrivateKey("ubuntu", "/home/ubuntu/.ssh/senior-design.pem", ssh.InsecureIgnoreHostKey())
```
Do not hardcode paths! Also, is there an alternative to "InsecureIgnoreHostKey" / is it actually insecure?

Others commented on scp.go:
```diff
@@ -0,0 +1,49 @@
+		Error.Println("Error creating ssh config", err)
+		return err
+	}
+	// For other authentication methods see ssh.ClientConfig and ssh.AuthMethod
```
Left a comment from the documentation here

Others commented on scp.go:
```diff
@@ -0,0 +1,49 @@
+	// For other authentication methods see ssh.ClientConfig and ssh.AuthMethod
+
+	// Create a new SCP client
+	client := scp.NewClientWithTimeout("ec2-54-211-200-158.compute-1.amazonaws.com:22", &clientConfig, time.Duration(100000000000))
```
Do not hardcode URLs

Others commented on scp.go:
```diff
@@ -0,0 +1,49 @@
+	// Finaly, copy the file over
+	// Usage: CopyFile(fileReader, remotePath, permission)
+	Info.Println("Copying " + tarName)
+	return client.CopyFile(f, dest, "0655")
```
Why `0655`? Needs a comment

hjaensch7 replied:
https://github.com/bramvdbogaerde/go-scp
This is the octal used in the example from the scp package. I can change it to something different if you want. Or maybe comment a small description of the permissions.

Others replied:
`0664` please, and add a comment that it means: read/write for owner/group, and then read only for everyone else

Others reviewed (changes requested):
Needs a lot more cleanup before being merged. I also improved the linter separately, so try and conform to that as well

Others committed:
Merge branch 'master' into dev

Others commented:
Reposting linter log here:
```
INFO [config_reader] Config search paths: [./ /root/project /root /] 
INFO [config_reader] Used config file .golangci.yml 
INFO [lintersdb] Active 29 linters: [deadcode depguard dupl errcheck goconst gocritic gocyclo gofmt goimports golint gosec gosimple govet ineffassign interfacer lll maligned misspell nakedret prealloc scopelint staticcheck structcheck stylecheck typecheck unconvert unparam unused varcheck] 
INFO [lintersdb] Optimized sublinters [staticcheck gosimple unused stylecheck] into metalinter megacheck 
INFO [loader] Go packages loading at mode load deps types and syntax took 30.02861691s 
INFO [loader] SSA repr building timing: packages building 26.109362ms, total 423.525568ms 
activate.go:11:2: don't use underscores in Go names; const worker_url should be workerURL (golint)
	worker_url = "http://ec2-54-211-200-158.compute-1.amazonaws.com/meta/activate"
	^
activate.go:14:6: don't use underscores in Go names; type dev_id should be devID (golint)
type dev_id struct {
     ^
activate.go:21:2: struct field `Id` should be `ID` (golint)
	Id               dev_id `json:"id"`
	^
activate.go:22:2: don't use underscores in Go names; struct field Executable_file should be ExecutableFile (golint)
	Executable_file  string `json:"executable_file"`
	^
activate.go:23:2: don't use underscores in Go names; struct field Execution_method should be ExecutionMethod (golint)
	Execution_method string `json:"execution_method"`
	^
activate.go:27:53: don't use underscores in Go names; func parameter execution_method should be executionMethod (golint)
func createActivateBody(dev dev_id, tarPath string, execution_method string) ([]byte, error) {
                                                    ^
main.go:44:3: don't use underscores in Go names; var `tar_name` should be `tarName` (golint)
		tar_name := guuid.New()
		^
activate.go:33:33: `activateWorker` - `worker` is unused (unparam)
func activateWorker(dev dev_id, worker string, tarPath string, tarName string) error {
                                ^
log.go:16:2: `setLogStreams` - `warningHandle` is unused (unparam)
	warningHandle io.Writer,
	^
log.go:17:2: `setLogStreams` - `errorHandle` is unused (unparam)
	errorHandle io.Writer) {
	^
main.go:27:12: declaration of "err" shadows declaration at main.go:21 (govet)
		payload, err := hook.Parse(r, github.PushEvent)
		         ^
main.go:68:3: assignOp: replace `tarNameExt = tarNameExt + ".gz"` with `tarNameExt += ".gz"` (gocritic)
		tarNameExt = tarNameExt + ".gz"
		^
main.go:4: File is not `goimports`-ed (goimports)
	guuid "github.com/google/uuid"
	"github.com/hashicorp/go-getter"
	"gopkg.in/go-playground/webhooks.v5/github"
scp.go:3: File is not `goimports`-ed (goimports)
import (
scp.go:45:5: `Finaly` is a misspelling of `Finally` (misspell)
	// Finaly, copy the file over
	   ^
scp.go:14: G106: Use of ssh InsecureIgnoreHostKey should be audited (gosec)
	clientConfig, err := auth.PrivateKey("ubuntu", "/home/ubuntu/.ssh/senior-design.pem", ssh.InsecureIgnoreHostKey())
INFO [runner] worker.2 took 335.690539ms with stages: unparam: 132.206229ms, interfacer: 85.736866ms, misspell: 56.601216ms, golint: 33.445058ms, goimports: 9.203422ms, dupl: 3.893137ms, gosec: 3.204555ms, gocritic: 2.946579ms, govet: 2.810124ms, gofmt: 2.549477ms, ineffassign: 1.01399ms, unconvert: 826.412µs, varcheck: 249.732µs, errcheck: 185.214µs, deadcode: 178.054µs, lll: 149.94µs, structcheck: 101.663µs, scopelint: 68.712µs, gocyclo: 61.114µs, prealloc: 50.715µs, goconst: 49.623µs, maligned: 45.904µs, nakedret: 29.268µs, depguard: 5.204µs, typecheck: 1.796µs 
scp.go:22: line is 128 characters (lll)
	client := scp.NewClientWithTimeout("ec2-54-211-200-158.compute-1.amazonaws.com:22", &clientConfig, time.Duration(100000000000))
INFO [runner] worker.1 took 2.502366706s with stages: megacheck: 2.502360506s 
INFO [runner] Workers idle times: #2: 2.154875866s 
INFO [runner] Issues before processing: 31, after processing: 17 
INFO [runner] processing took 6.382584ms with stages: exclude: 3.34151ms, identifier_marker: 1.385643ms, cgo: 324.999µs, skip_dirs: 251.608µs, autogenerated_exclude: 213.949µs, source_code: 199.558µs, path_prettifier: 156.727µs, nolint: 143.614µs, replacement_builder: 101.155µs, uniq_by_line: 97.247µs, max_same_issues: 72.249µs, max_from_linter: 30.661µs, max_per_file_from_linter: 29.821µs, path_shortener: 27.243µs, exclude-rules: 3.474µs, diff: 1.772µs, skip_files: 1.354µs 
INFO Memory: 304 samples, avg is 114.5MB, max is 1148.1MB 
INFO Execution took 33.182654512s                 

Exited with code exit status 1
```

hjaensch7 committed:
PR changes, relative paths, and .env config

hjaensch7 committed:
Updated SCP.go to use env variable path

hjaensch7 committed:
SCP works using .env variable

This commit includes linter corrections and minor PR changes

hjaensch7 committed:
Run gofmt -w on scp.go and main.go

hjaensch7 committed:
Fixed activateWorker and ran goimports

hjaensch7 committed:
Remove executable and add it to .gitignote :-(

hjaensch7 committed:
Moved github push event handler out of main

hjaensch7 committed:
Server receives status request and push events

hjaensch7 committed:
Get status from both workers and combine

Others commented on .env:
```diff
@@ -0,0 +1,5 @@
+CI_PORT="0.0.0.0:81"
+WEBSITE_PORT="0.0.0.0:80"
```
These should come from the `--development` flag, not the environment

Others commented on .env:
```diff
@@ -0,0 +1,5 @@
+CI_PORT="0.0.0.0:81"
+WEBSITE_PORT="0.0.0.0:80"
+WORKER1="v9_w1.patcody.io"
+WORKER2="v9_w2.patcody.io"
```
These need to be one env variable. Should be:

```
V9_WORKERS="v9_w1.patcody.io; v9_w2.patcody.io"
```

Others commented on .golangci.yml:
```diff
@@ -18,4 +18,4 @@ linters:
     - gochecknoinits
     - gochecknoglobals
-
+    - gosec
```
Newline at end of file

Others commented on activate.go:
```diff
@@ -0,0 +1,54 @@
+	}
+	// Make activate post request
+	workerURL = "http://" + workerURL
+	workerURL += "/meta/activate"
```
Why two lines?

Others commented on activate.go:
```diff
@@ -0,0 +1,54 @@
+	}
+
+	Info.Println("Response from worker:", string(respBody))
+	return err
```
`return nil` is clearer

Others commented on docker.go:
```diff
@@ -0,0 +1,48 @@
+
+// Build Docker Image Based on Dockerfile
+func buildImageFromDockerfile(tarName string) error {
+	cmd := exec.Command("docker", "build", "-t", tarName, "./temp_repo")
```
`./temp_repo` is hardcoded here

Others replied:
You didn't actually switch to using the global here

Others commented on docker.go:
```diff
@@ -0,0 +1,48 @@
+	if err != nil {
+		Error.Println("Error building tar from image", err)
+		return ""
+	}
```
Add an `error` return from this function, then propagate errors upward

Others commented on docker.go:
```diff
@@ -0,0 +1,48 @@
+	if err != nil {
+		Error.Println("Failure to gzip")
+		return ""
+	}
```
Add an error return from this function, then propagate errors upward

Others commented on main.go:
```diff
@@ -1,82 +1,74 @@
+	//Load .env file
+	if err := godotenv.Load(); err != nil {
+		Error.Print("No .env file found")
+	}
```
You don't want to do this. This prevents the programmer for overriding the `.env` file. Instead they need to source it before running you

Others commented on main.go:
```diff
@@ -1,82 +1,74 @@
-	return "git::" + p.Repository.URL
-}
+// Get env variables
+func getEnvVariables() (string, string, []string) {
```
This function should have an error return value as well

Others commented on statusHandler.go:
```diff
@@ -0,0 +1,45 @@
+
+import (
+	"encoding/json"
+	//	"fmt"
```
???

hjaensch7 replied:
this is bad, I'm sorry

hjaensch7 committed:
Handle json.encode error

Others commented on .env:
```diff
@@ -0,0 +1,5 @@
+WEBSITE_PORT="0.0.0.0:80"
+WORKER1="v9_w1.patcody.io"
+WORKER2="v9_w2.patcody.io"
+
```
Don't include `.env` in the push

hjaensch7 committed:
Multiple worker deployment

hjaensch7 committed:
Removed and replaced .env

hjaensch7 committed:
Improved error propagation

hjaensch7 committed:
Fixing linter issues

Others committed:
Add CORS workaround

Others commented on activate.go:
```diff
@@ -0,0 +1,54 @@
+// Activate worker
+func activateWorker(dev devID, workerURL string, tarPath string) error {
+	// Marshal information into json body
+	body, err := createActivateBody(dev, tarPath, "docker-archive")
+	if err != nil {
+		Error.Println("Failed to create activation body", err)
+		return err
+	}
+
+	// Make activate post request
+	workerURL = "http://" + workerURL + "/meta/activate"
+	resp, err := http.Post(workerURL, "application/json", bytes.NewBuffer(body))
+	if err != nil {
+		Error.Println("Failed to post", err)
+		return err
+	}
```
Nit: In the future we probably want to create a "V9Worker" object to manage this stuff
(add a todo, don't bother adding to PR)

hjaensch7 replied:
what stuff in particular? the error handling? making  the post?

Others replied:
Error handling, url manipulation, query serialization ect.

Others replied:
Error handling, url manipulation, query serialization ect.

Others commented on activate.go:
```diff
@@ -0,0 +1,54 @@
+	User string `json:"user"`
+	Repo string `json:"repo"`
+	Hash string `json:"hash"`
+}
```
Used for both activate and deactivate. Factor this out to another file.

Others commented on env.sh:
```diff
@@ -0,0 +1,5 @@
+export CI_PORT="0.0.0.0:81"
+export WEBSITE_PORT="0.0.0.0:80"
+export WORKER1="v9_w1.patcody.io"
+export WORKER2="v9_w2.patcody.io"
```
This file should not be pushed, please delete

Others replied:
(We might need a private repo for shared configuration)

hjaensch7 replied:
got it will delete

Others commented on main.go:
```diff
@@ -1,82 +1,61 @@
-	path = "/payload"
-	port = ":3069"
+	CIPort      = "0.0.0.0:81"
+	websitePort = "0.0.0.0:80"
```
You need to check for the `--development` flag to decide on what port to use

Others commented on main.go:
```diff
@@ -1,82 +1,61 @@
 	}
-	return nil
+	workerArr[1] = worker
+	return workerArr, nil
```
Use the `V9_WORKERS` instead of `WORKER1` and `WORKER2`

Others commented on pushHandler.go:
```diff
@@ -0,0 +1,95 @@
+)
+
+type pushHandler struct {
+	worker  []string
```
`worker` => `workers`

Others commented on pushHandler.go:
```diff
@@ -0,0 +1,95 @@
+	// Send .tar to worker
+	Info.Println("SCP tar to worker...")
+	source := "./" + tarNameExt
+	destination := "/home/ubuntu/" + tarNameExt
```
Nit: Perhaps we should be sending to a location better than `/home/ubuntu`
Add a todo comment, don't bother implementing for this PR

hjaensch7 replied:
got it

Others commented on pushHandler.go:
```diff
@@ -0,0 +1,95 @@
+	// Activate worker
+	user := push.Repository.Owner.Login
+	repo := push.Repository.Name
+	dev := devID{user, repo, "test_hash"}
```
Nit: We need to use the actual commit hash 
Add a TODO, don't bother for this PR

hjaensch7 replied:
got it

Others commented on pushHandler.go:
```diff
@@ -0,0 +1,95 @@
+	counter int
+}
+
+func (h *pushHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```
Nit: This function is buggy if we get a push before the previous deployment finishes
Add a TODO comment, don't bother to fix for this PR

Others commented on run.sh:
```diff
@@ -0,0 +1 @@
+sudo sh -c ". ./env.sh;./v9_deployment_manager"
```
Also should not be pushed

Others commented on scp.go:
```diff
@@ -0,0 +1,50 @@
+func scpToWorker(workerURL string, source string, dest string, tarName string) error {
+	// Use SSH key authentication from the auth package
+	// we ignore the host key in this example, please change this if you use this library
+	clientConfig, err := auth.PrivateKey("ubuntu", "/home/ubuntu/.ssh/senior-design.pem", ssh.InsecureIgnoreHostKey())
```
Add a TODO to do something smarter about the host key

Others commented on scp.go:
```diff
@@ -0,0 +1,50 @@
+
+	// Create a new SCP client
+	workerURL += ":22"
+	client := scp.NewClientWithTimeout(workerURL, &clientConfig, time.Duration(100000000000))
```
Switch to 5 minutes timeout, and make it more readable

Others commented on statusHandler.go:
```diff
@@ -0,0 +1,53 @@
+		// Parse response
+		respBody, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			Error.Println("Failure to parse response from worker", err)
```
"read response" not parse response, right?

Others reviewed (changes requested):
Much closer, but has a few configuration problems

hjaensch7 committed:
Handle pr comments, remove env.sh and run.sh

hjaensch7 committed:
Merge branch 'dev' of github.com:velocity-9/v9_deployment_manager into dev

hjaensch7 committed:
Added development flag, and changed worker url format

Others commented on consts.go:
```diff
@@ -0,0 +1,5 @@
+package main
+
+const (
+	tempRepoPath = "./temp_repo"
+)
```
We want request concurrency. Using one file path is racy. You need to create a new random temp folder for each request (there probably is a library for this)

Maybe we can do this with the other fix later

hjaensch7 replied:
will add this to the issue

Others commented on main.go:
```diff
@@ -1,82 +1,62 @@
+	if len(os.Args) > 1 && os.Args[1] == "--development" {
+		CIPort = ":3081"
+		websitePort = ":3080"
+	}
```
try if `os.Args` contains ...

hjaensch7 replied:
there is no contains method but I guess I could just write one.

Others replied:
If I do `./deployment_manager random-junk --development`, then it shouldn't just silently ignore the development flag

hjaensch7 replied:
that makes sense, I'll add a contains check

Others commented on main.go:
```diff
@@ -1,82 +1,62 @@
-	return nil
+	workerArr := strings.Split(workerString, ";")
+	return workerArr, nil
 }
```
Follow Pat's convention: `V9_WORKERS`

Others commented on main.go:
```diff
@@ -1,82 +1,62 @@
-	if err != nil {
-		return err
+// Get env variables
+func getEnvVariables() ([]string, error) {
```
rename: getWorkers

Others commented on pushHandler.go:
```diff
@@ -0,0 +1,91 @@
+	worker := h.workers[h.counter]
+	h.counter = (h.counter + 1) % len(h.workers)
+	// Setup github webhook
+	hook, githubErr := github.New(github.Options.Secret("thespeedeq"))
```
Load from env variable

Others commented on scp.go:
```diff
@@ -0,0 +1,52 @@
+	// Create a new SCP client
+	workerURL += ":22"
+	// Timeout after 5 minutes
+	client := scp.NewClientWithTimeout(workerURL, &clientConfig, time.Duration(300000000000))
```
there has to be a clearer way to write `5 minutes`

hjaensch7 replied:
I could define a minute as 60000000000 nanoseconds somewhere else and then write 5*minute if that's cleaner

Others replied:
Either use `time.ParseDuration` or `5 * time.Minute`

hjaensch7 replied:
will do

Others commented on scp.go:
```diff
@@ -0,0 +1,52 @@
+func scpToWorker(workerURL string, source string, dest string, tarName string) error {
+	// Use SSH key authentication from the auth package
+	// we ignore the host key in this example, please change this if you use this library
+	clientConfig, err := auth.PrivateKey("ubuntu", "/home/ubuntu/.ssh/senior-design.pem", ssh.InsecureIgnoreHostKey())
```
Make an issue for this line -- there are significant problems with it

hjaensch7 replied:
will do

Others commented on scp.go:
```diff
@@ -0,0 +1,52 @@
+	}
+
+	// Close client connection after the file has been copied
+	defer client.Close()
```
Should go above the file opening

Others commented on statusHandler.go:
```diff
@@ -0,0 +1,53 @@
+	}
+	Info.Println("Sending Status...")
+
+	// FIXME: This CORS workaround cannot be in the final version
+	w.Header().Set("Access-Control-Allow-Origin", "*")
```
Add issue for this

Others commented on docker.go:
```diff
@@ -0,0 +1,48 @@
+	Info.Println("Gzipping tar...")
+	err = gzipTar(tarNameExt)
+	if err != nil {
+		Error.Println("Failure to gzip")
```
Print the `err`

hjaensch7 replied:
should they be printed here or after they've been returned and checked or both

Others commented on pushHandler.go:
```diff
@@ -0,0 +1,91 @@
+	Info.Println("Building and zipping tar...")
+	tarNameExt, err := buildAndZipTar(tarName)
+	if err != nil {
+		Error.Println("Failed to build and compress tar")
```
Print the err

Others commented on statusHandler.go:
```diff
@@ -0,0 +1,53 @@
+		if err != nil {
+			Error.Println("Failed to get status", err)
+			return
+		}
```
Probably want to continue instead of returning

Others commented on statusHandler.go:
```diff
@@ -0,0 +1,53 @@
+func (h *statusHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Make call out to get worker 2 status
+	Info.Println("Getting Worker Status...")
+	var allStat = allStatus{}
```
why not just use a slice?  No need for this new type

hjaensch7 replied:
agree, this was dumb

Others reviewed (changes requested):
Good work on this. Very close!

In future definitely need to keep these shorter, it's somewhat overwhelming to review

Others commented:
@hjaensch7  You've marked a lot of my comments as resolved? Can you push the relevant code?

In general, it's kinda confusing to mark so much as resolved without a contemporaneous push--it's difficult as a reviewer to know what comments are actually resolved.

I've addressed a couple more follow ups. We're almost there on this! Let's push on it!

hjaensch7 committed:
Update error propegation and started reading secret from file, also addressed pr comments marked resolved

hjaensch7 committed:
Preallocated status array based on number of workers

hjaensch7 committed:
Removed type from statushandler.go:21

hjaensch7 committed:
Updated go version in circle ci

hjaensch7 committed:
Reversed circleci go version upgrade

hjaensch7 committed:
Updated go version in circle ci again

hjaensch7 committed:
Updated go, added go.mod and go.sum

hjaensch7 committed:
Added golang install to circleci config

hjaensch7 committed:
Revert previous commit

hjaensch7 committed:
Attempt to specify circlici linter version

hjaensch7 committed:
Attempt to specify circlici linter version 2

hjaensch7 committed:
Pulled deactivate components loop into deactivate.go

hjaensch7 committed:
Attempt to turn off funlen check

hjaensch7 committed:
Added contains function for development flag check

hjaensch7 committed:
Changed time to be more readable

Others commented on .gitignore:
```diff
@@ -10,3 +10,8 @@
 # Output of the go coverage tool, specifically when used with LiteIDE
 *.out
+
+v9_deployment_manager
+env.sh
+run.sh
```
Nit: Needs a comment

Others commented on .golangci.yml:
```diff
@@ -18,4 +18,6 @@ linters:
     - gochecknoinits
     - gochecknoglobals
+    - gosec
+    - funlen
```
ooof. Please make issue to re-enable

Others commented on activate.go:
```diff
@@ -0,0 +1,48 @@
+		return err
+	}
+
+	defer resp.Body.Close()
```
Nit: spacing off again

Others commented on deactivate.go:
```diff
@@ -0,0 +1,58 @@
+		return err
+	}
+
+	defer resp.Body.Close()
```
Nit: spacing weird

Others commented on deactivate.go:
```diff
@@ -0,0 +1,58 @@
+}
+
+// Deactivate individual component
+func deactivateIndividualComponent(dev devID, workerURL string) error {
```
Nit: Bad name. It disables an component on a single worker

Others reviewed (approved):
I'm approving this now. Please squash and merge, then follow up with a PR that fixes the Nits. Then cover each Github issue in small PRs. This PR is too big
